File: Fundamentals_of_Software_Design/1_Basic_skills/Удаление пробелов в начале и конце строки/main.c
#include <stdio.h>
#include <ctype.h>
#include <string.h>

void strtrim(char *string) {
    if (!string || *string == '\0') return;

    // Указатель на начало строки без начальных пробелов
    char *start = string;
    while (isspace((unsigned char)*start)) {
        start++;
    }

    // Указатель на конец строки без конечных пробелов
    char *end = string + strlen(string) - 1;
    while (end > start && isspace((unsigned char)*end)) {
        *end = '\0'; // Обнуляем символы пробелов
        end--;
    }

    // Сдвигаем оставшуюся часть строки к началу
    if (start != string) {
        memmove(string, start, end - start + 2);
    }
}

int main() {
    char buf[] = "   abc   ";
    printf("<%s>\n", buf); // <   abc   >
    strtrim(buf);
    printf("<%s>\n", buf); // <abc>
    return 0;
}

File: Fundamentals_of_Software_Design/1_Basic_skills/Удаление пробелов в начале и конце строки/main.py
def strtrim(string):
    """
    Удаляет пробелы в начале и конце строки.
    Изменяет строку, если она передана как изменяемый объект (например, list).
    """
    if not string:
        return ""

    # Удаление пробелов в начале
    start = 0
    while start < len(string) and string[start].isspace():
        start += 1

    # Удаление пробелов в конце
    end = len(string) - 1
    while end >= start and string[end].isspace():
        end -= 1

    # Возврат среза строки
    return string[start:end + 1]


# Пример использования
buf = "   abc   "
print(f"<{buf}>")  # <   abc   >
trimmed = strtrim(buf)
print(f"<{trimmed}>")  # <abc>
